kubectl: Add proxy --unix-socket=/file/path option

stefwalter opened the pull request:
Proxies on a TCP port are accessible outside the current security
context (eg: uid). Add support for having the proxy listen on a
unix socket, which has permissions applied to it.

We make sure the socket starts its life only accessible by the
current user using Umask.

This is useful for applications like Cockpit and other tools which
want the help of kubectl to handle authentication, configuration and
transport security, but also want to not make that accessible to
all users on a multi-user system.

stefwalter committed:
kubectl: Add proxy --unix-socket=/file/path option

Proxies on a TCP port are accessible outside the current security
context (eg: uid). Add support for having the proxy listen on a
unix socket, which has permissions applied to it.

We make sure the socket starts its life only accessible by the
current user using Umask.

This is useful for applications like Cockpit and other tools which
want the help of kubectl to handle authentication, configuration and
transport security, but also want to not make that accessible to
all users on a multi-user system.

stefwalter commented:
@eparis When you have time, is this something you'd be interested in reviewing?

zmerlynn commented:
This looks like something we could wait until after 1.0 to take, but punting over to @jlowdermilk.

eparis commented:
It LGTM and we would certainly like to see it in 1.0. Although would understand a rejection. It's obviously safe from a 'breaks other stuff' PoV

jlowdermilk commented on pkg/kubectl/cmd/proxy.go:
```diff
@@ -70,13 +70,19 @@ The above lets you 'curl localhost:8001/custom/api/v1/pods'
+	path := cmdutil.GetFlagString(cmd, "unix")
 	port := cmdutil.GetFlagInt(cmd, "port")
-	fmt.Fprintf(out, "Starting to serve on localhost:%d", port)
+	if path == "" {
```
Nit: move print statements down before call to `server.Serve` so you don't have to do this same check twice.

jlowdermilk commented:
Small nit otherwise lgtm. Agree that it's a low risk feature. Deferring to @bgrant0607 to make the call whether it's v1.0 or post.

bgrant0607 commented on pkg/kubectl/cmd/proxy.go:
```diff
@@ -70,13 +70,19 @@ The above lets you 'curl localhost:8001/custom/api/v1/pods'
 	cmd.Flags().String("reject-methods", kubectl.DefaultMethodRejectRE, "Regular expression for HTTP methods that the proxy should reject.")
 	cmd.Flags().IntP("port", "p", 8001, "The port on which to run the proxy.")
-	cmd.Flags().Bool("disable-filter", false, "If true, disable request filtering in the proxy. This is dangerous, and can leave you vulnerable to XSRF attacks.  Use with caution.")
+	cmd.Flags().StringP("unix", "u", "", "Unix socket on which to run the proxy.")
```
--unix is too obscure. How about --unix-socket

bgrant0607 commented on pkg/kubectl/cmd/proxy.go:
```diff
@@ -70,13 +70,19 @@ The above lets you 'curl localhost:8001/custom/api/v1/pods'
+	if path == "" {
+		fmt.Fprintf(out, "Starting to serve on localhost:%d\n", port)
+	} else {
+		fmt.Fprintf(out, "Starting to serve on %s\n", path)
```
Please generate an error if port is set to a non-default value and --unix-socket is also specified.

bgrant0607 commented:
I'm fine with this particular change, but I need to stop even looking at PRs containing anything other than critical bug fixes very, very soon.

Install pre-submit hooks.
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/development.md#hooks

```
./hack/verify-gofmt.sh
!!! 'gofmt -s' needs to be run on the following files:
./pkg/kubectl/proxy_server.go
```

stefwalter commented:
Updated.
 * Changed to ```--unix-socket```
 * Check for both ```--unix-socket``` and ```--port```
 * Updated docs
 * Fixed gofmt
 * Moved print statements.

stefwalter commented:
Fixed issue with warning about filter for unix sockets.

jlowdermilk commented:
LGTM

stefwalter commented:
Rebased commit from #10680 onto this branch, as suggested by @bgrant0607

smarterclayton commented:
Code LGTM, but I think this deserves some sort of e2e test.

stefwalter commented:
@eparis Where would I put an e2e test? And how can I run these tests. They fall over when I try to poke them in an ignorant manner:

```
$ sudo ./_output/local/bin/linux/amd64/e2e --provider=local
...
FAIL! -- 1 Passed | 43 Failed | 1 Pending | 5 Skipped F0702 22:21:10.758207   25200 driver.go:94] At least one test failed
```

smarterclayton commented:
development.md
<https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/development.md#end-to-end-tests>

On Thu, Jul 2, 2015 at 4:21 PM, Stef Walter <notifications@github.com>
wrote:

> @eparis <https://github.com/eparis> Where would I put an e2e test? And
> how can I run these tests. They fall over when I try to poke them in an
> ignorant manner:
>
> $ sudo ./_output/local/bin/linux/amd64/e2e --provider=local
> ...
> FAIL! -- 1 Passed | 43 Failed | 1 Pending | 5 Skipped F0702 22:21:10.758207   25200 driver.go:94] At least one test failed
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10608#issuecomment-118154883>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

stefwalter committed:
util: Add util.Umask wrapper

This function compiles to a stub on Windows.

bprashanth commented:
@stefwalter are we blocking this on an e2e?

stefwalter committed:
test: End to end test for kubectl proxy --unix-socket

bgrant0607 commented:
Re. e2e, see also #10764. 

This needs to wait until after 1.0.

mikedanese commented:
@stefwalter needs rebase.

alex-mohr commented:
@stefwalter Thanks for the PR -- can you please rebase, and I'll try to merge quickly after that?

stefwalter commented:
@alex-mohr Thanks, rebased.

mikedanese commented:
needs a gofmt as well.

stefwalter commented:
> needs a gofmt as well.

Done and generated docs.

mikedanese commented:
Tests are also not compiling. 
```
make test
...
# github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl
_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/proxy_server_test.go:290: too many arguments in call to NewProxyServer
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl [build failed]
```

stefwalter commented:
Fixed ```make test```

jlowdermilk commented:
@stefwalter, what happened to the e2e test that was included with this change? Should be pretty easy to add a test for `--unix-socket` under [Describe("Proxy server", ...](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/kubectl.go#L529)

stefwalter commented:
> what happened to the e2e test that was included with this change? Should be pretty easy to add a test for --unix-socket under Describe("Proxy server", ...

Seems more complicated due to the use of a Unix socket with net.http ... but I can give it a shot.

stefwalter commented:
Ok, I think this is ready.

jlowdermilk commented on test/e2e/kubectl.go:
```diff
@@ -546,6 +548,35 @@ var _ = Describe("Kubectl client", func() {
 	})
 
+	Describe("Proxy unix socket", func() {
+		It("should support --unix-socket=/path", func() {
```
Nit: I would move this above under the existing `Describe()` for proxy server, since it is testing a different option of the same command.

jlowdermilk commented:
Minor nit, LGTM. Thanks for adding the e2e test.
